middleware/proxy/proxy.go

```diff
@@ -2,83 +2,181 @@
 
 import (
+	"errors"
+	"github.com/mholt/caddy/middleware"
+	"net"
 	"net/http"
-	"net/http/httputil"
 	"net/url"
+	"regexp"
 	"strings"
-
-	"github.com/mholt/caddy/middleware"
+	"sync/atomic"
+	"time"
 )
 
+var errUnreachable = errors.New("Unreachable backend")
+
 // Proxy represents a middleware instance that can proxy requests.
 type Proxy struct {
-	Next  middleware.Handler
-	Rules []Rule
+	Next      middleware.Handler
+	Upstreams []Upstream
 }
 
-// ServeHTTP satisfies the middleware.Handler interface.
-func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
+// An upstream manages a pool of proxy upstream hosts. Select should return a
+// suitable upstream host, or nil if no such hosts are available.
+type Upstream interface {
+	// The path this upstream host should be routed on
+	From() string
+	// Selects an upstream host to be routed to.
+	Select() *UpstreamHost
+}
 
-	for _, rule := range p.Rules {
-		if middleware.Path(r.URL.Path).Matches(rule.From) {
-			var base string
-
-			if strings.HasPrefix(rule.To, "http") { // includes https
-				// destination includes a scheme! no need to guess
-				base = rule.To
-			} else {
-				// no scheme specified; assume same as request
-				var scheme string
-				if r.TLS == nil {
-					scheme = "http"
-				} else {
-					scheme = "https"
-				}
-				base = scheme + "://" + rule.To
-			}
+type UpstreamHostDownFunc func(*UpstreamHost) bool
 
-			baseUrl, err := url.Parse(base)
-			if err != nil {
-				return http.StatusInternalServerError, err
-			}
-			r.Host = baseUrl.Host
+// An UpstreamHost represents a single proxy upstream
+type UpstreamHost struct {
+	Name         string
+	ReverseProxy *ReverseProxy
+	Conns        int64
+	Fails        int32
+	FailTimeout  time.Duration
+	Unhealthy    bool
+	ExtraHeaders http.Header
+	CheckDown    UpstreamHostDownFunc
+}
 
-			// TODO: Construct this before; not during every request, if possible
-			proxy := httputil.NewSingleHostReverseProxy(baseUrl)
-			proxy.ServeHTTP(w, r)
-			return 0, nil
-		}
+func (uh *UpstreamHost) Down() bool {
+	if uh.CheckDown == nil {
+		// Default settings
+		return uh.Unhealthy || uh.Fails > 0
 	}
+	return uh.CheckDown(uh)
+}
 
-	return p.Next.ServeHTTP(w, r)
+//https://github.com/mgutz/str
+var tRe = regexp.MustCompile(`([\-\[\]()*\s])`)
+var tRe2 = regexp.MustCompile(`\$`)
+var openDelim = tRe2.ReplaceAllString(tRe.ReplaceAllString("{{", "\\$1"), "\\$")
+var closDelim = tRe2.ReplaceAllString(tRe.ReplaceAllString("}}", "\\$1"), "\\$")
+var templateDelim = regexp.MustCompile(openDelim + `(.+?)` + closDelim)
+
+type requestVars struct {
+	Host         string
+	RemoteIp     string
+	Scheme       string
+	Upstream     string
+	UpstreamHost string
 }
 
-// New creates a new instance of proxy middleware.
-func New(c middleware.Controller) (middleware.Middleware, error) {
-	rules, err := parse(c)
-	if err != nil {
-		return nil, err
+func templateWithDelimiters(s string, vars requestVars) string {
```

Comment:
I hadn't seen it and I honestly wasn't sure how to best continue with string interpolation here which is why I went the route I did (I originally tried to use `text/template`). Using the already supplied `Replacer` is much cleaner IMO.